casanovo/utils.py

```diff
@@ -66,3 +75,221 @@ def split_version(version: str) -> Tuple[str, str, str]:
+
+    Parameters
+    ----------
+    peptide_lengths: np.ndarray
+        Numpy array containing the length of each sequence
+
+    Returns
+    -------
+    peptide_length_hist_matrix: np.ndarray
+        A matrix of size (n x 2) where each row vector is
+        (peptide_length, frequency)
+    """
+    frequncies = np.bincount(peptide_lengths).reshape((-1, 1))
+    bins = np.arange(len(frequncies), dtype=int).reshape((-1, 1))
+    bin_freqs = np.hstack((bins, frequncies))
+    return bin_freqs[bin_freqs[:, 1] != 0]
+
+
+def get_peptide_lengths(results_table: pd.DataFrame) -> np.ndarray:
+    """
+    Get a numpy array containing the length of each peptide sequence
+
+    Parameters
+    ----------
+    results_table: pd.DataFrame
+        Parsed spectrum match table
+
+    Returns
+    -------
+    sequence_lengths: np.ndarray
+        Numpy array containing the length of each sequence, listed in the
+        same order that the sequences are provided in.
+    """
+    # Mass modifications do not contribute to sequence length
+    # If PTMs ae represented in ProForma notation this filtering operation
+    # needs to be reimplemented
+    return (
+        results_table["sequence"]
+        .str.replace(r"[^a-zA-Z]", "", regex=True)
+        .apply(len)
+    )
+
+
+def get_report_dict(
+    results_table: pd.DataFrame, score_bins: List[float] = SCORE_BINS
+) -> Optional[Dict]:
+    """
+    Generate sequencing run report
+
+    Parameters
+    ----------
+    results_table: pd.DataFrame
+        Parsed spectrum match table
+    score_bins: List[float], Optional
+        Confidence scores for creating confidence CMF, see get_score_bins
+
+    Returns
+    -------
+    report_gen: Dict
+        Generated report represented as a dictionary, or None if no
+        sequencing predictions were logged
+    """
+    if results_table.empty:
+        return None
+
+    peptide_lengths = get_peptide_lengths(results_table)
+    min_length, med_length, max_length = np.quantile(
+        peptide_lengths, [0, 0.5, 1]
+    )
+    return {
+        "num_spectra": len(results_table),
+        "score_bins": get_score_bins(results_table, score_bins),
+        "max_sequence_length": max_length,
+        "min_sequence_length": min_length,
+        "median_sequence_length": med_length,
+        "peptide_length_histogram": get_peptide_length_hist(peptide_lengths),
+    }
+
+
+def log_run_report(
+    start_time: Optional[int] = None, end_time: Optional[int] = None
+) -> None:
+    """
+    Log general run report
+
+    Parameters
+    ----------
+    start_time : Optional[int], default=None
+        The start time of the sequencing run in seconds since the epoch.
+    end_time : Optional[int], default=None
+        The end time of the sequencing run in seconds since the epoch.
+    """
+    logger.info("======= End of Run Report =======")
+    if (start_time is not None) and (end_time is not None):
+        end_time = time.time() if end_time is None else end_time
```

Comment:
**nitpick:** `end_time` will never be `None` based on the `if` the preceding line.